### ISSUE TITLE
Training with Larger Image results in 0 P, R, mAP 

### ISSUE THREAD

**Issue by ashwin-999** [ultralytics/yolov5#12837]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi, 
I am running into a situation where training with 640x640 seem to work fine and seem to get reasonable results. However when I switch to 960x960, things seem totally off. The P, R, mAPs for the first many epochs is 0. It's the exact same command and code I'm running with an additional flag `--imgsz 960`. Is this behavior expected? Else, what could be happening in training that I should address

**Comment by glenn-jocher** [comment]
@ashwin-999 hi there! 😊

Experiencing a sudden drop in performance metrics (P, R, mAP) when switching image sizes from 640x640 to 960x960 is unusual but could stem from several factors. Here are a couple of quick points to check:
- **Batch Size:** Adjusting the image size impacts GPU memory usage, potentially leading to suboptimal batch sizes. Ensure your batch size is still suitable for the larger image dimension.
- **Learning Rate:** Larger images might require tweaking the learning rate. Consider adjusting it slightly.
- **Augmentations:** Ensure your data augmentations are appropriate for the larger resolution. Sometimes, augmentations that work well for smaller images may not scale effectively.
- **Convergence Time:** Training with larger images generally takes longer. The metrics might improve if you allow more epochs for the model to converge.

Remember, modeling is often iterative and requires experimentation. Adjust parameters gradually and monitor the impact.

Happy training! 🚀

**Comment by ashwin-999** [comment]
Hey @glenn-jocher thanks for your response. How important is it to change this if we move from 640 -> other image sizes??

`b.data[:, 4] += math.log(8 / (640 / s) ** 2)  # obj (8 objects per 640 image)`

Also, are there any studies on using different `anchor_t`? I seem to get 100% recall with `anchor_t=5` and Best Possible Recall drops to 0.48 with `anchor_t=4.5` before anchors are recomputed. Any thoughts how such a sharp change might be explained?

**Comment by glenn-jocher** [comment]
@ashwin-999 hi there! 👋

Switching from 640 to other image sizes does necessitate adjustments in the scaling of objects, as represented by the line of code you're referring to. The importance of this adjustment stems from ensuring that the objectness target sizes remain consistent relative to the new image size. This helps in maintaining or improving the model's detection capabilities across various dimensions.

Regarding the `anchor_t` parameter and its impact on recall, the sharp change you observed between `anchor_t=5` and `anchor_t=4.5` can be attributed to how anchor thresholds influence the matching process between ground truths and anchor boxes. A higher `anchor_t` value typically allows for a more generous match, potentially increasing recall but at the risk of reducing precision. The drastic change suggests a sensitive balance in your dataset between false positives and true positives at these thresholds.

Each dataset can behave differently, so these observations might not universally apply. Experimentation is key. Try visualizing your anchors and their match quality at different `anchor_t` values to better understand the underlying cause.

Happy experimenting! 🚀

**Comment by ashwin-999** [comment]
@glenn-jocher thanks for your insights. I am trying out with a large number of anchors while keeping `anchor_t=4.0`

Here is what I am seeing

with the default anchors the fit is bad, so the `autoanchor` recomputes new anchors. However what's concerning is the width I see on the newly computed anchors. (See below) Almost none of my objects have a width of 100+ but I get anchors with width in the 100 + range. But when I set anchors to 15 I'm forcing auto anchor to group differently. When I do this I see anchors that are much more representative of the object sizes in the images. 

I am a bit lost on how to explain this to myself. I would like to hear your thoughts 

Default:
```
Best Possible Recall: 0.375 with Default Anchors

AutoAnchor: 0.63 anchors/target, 0.375 Best Possible Recall (BPR). Anchors are a poor fit to dataset ⚠️, attempting to improve...
AutoAnchor: Running kmeans for 12 anchors on 73472 points...
AutoAnchor: Evolving anchors with Genetic Algorithm: fitness = 0.8918: 100%|██████████| 1000/1000 [00:02<00:00, 399.21it/s]
AutoAnchor: thr=0.25: 0.9950 best possible recall, 11.05 anchors past thr
AutoAnchor: n=12, img_size=640, metric_all=0.564/0.893-mean/best, past_thr=0.596-mean: 86,16, 153,16, 122,22, 189,14, 274,14, 251,16, 326,14, 385,14, 414,16, 489,14, 455,24, 325,37
Best Possible Recall with New Achors: 0.9949096441268921
AutoAnchor: Done ✅ (optional: update model *.yaml to use these anchors in the future)
AutoAnchor: Done ✅ (optional: update model *.yaml to use these anchors in the future)
```

With `anchors=15`
```
AutoAnchor: 1.56 anchors/target, 0.103 Best Possible Recall (BPR). Anchors are a poor fit to dataset ⚠️, attempting to improve...
AutoAnchor: Running kmeans for 45 anchors on 73472 points...
AutoAnchor: Evolving anchors with Genetic Algorithm: fitness = 0.9447: 100%|██████████| 1000/1000 [00:21<00:00, 47.21it/s]
AutoAnchor: thr=0.25: 0.9994 best possible recall, 39.87 anchors past thr
AutoAnchor: n=45, img_size=640, metric_all=0.562/0.945-mean/best, past_thr=0.612-mean: 45,16, 72,13, 68,19, 65,21, 104,14, 98,17, 140,14, 176,14, 208,13, 155,18, 140,20, 175,17, 234,13, 219,15, 217,16, 104,33, 255,14, 243,15, 272,13, 219,17, 301,13, 292,14, 171,24, 280,15, 274,16, 260,17, 238,19, 330,14, 358,13, 410,13, 376,14, 362,15, 326,17, 457,13, 424,14, 418,15, 387,17, 482,14, 452,17, 576,14, 420,21, 557,18, 488,25, 522,30, 419,43
Best Possible Recall with New Achors: 0.9994011521339417
AutoAnchor: Done ✅ (optional: update model *.yaml to use these anchors in the future)
AutoAnchor: Done ✅ (optional: update model *.yaml to use these anchors in the future)
```

**Comment by glenn-jocher** [comment]
@ashwin-999, hi there! 👋 It sounds like you're diving deep into the autoanchor settings - that's great!

When you increase the number of anchors (e.g., `anchors=15` leading to 45 anchors due to 3x multiplication), the AutoAnchor algorithm has more flexibility to fit the diverse sizes of objects within your dataset. This often leads to a better representation of your actual object sizes, as you've noticed with anchor sizes becoming more representative. 🎯

The appearance of unexpectedly large anchor widths in the default setting could be due to the algorithm trying to cover as much variance in object sizes with a limited number of anchors, leading to some compromises. Essentially, with more anchors, the algorithm can afford to "specialize" each anchor more finely to the distribution of sizes in your dataset.

If the new anchors with `anchors=15` significantly improve your Best Possible Recall and seem more representative of your dataset, it could be beneficial to update your model's YAML file with these anchors for future training. This customization tailors the model more closely to the specifics of your data, potentially enhancing its performance.

Keep experimenting and observing the effects! 🚀
